src/devices/cpu/mpk1839/kl1839vm1.cpp

```diff
@@ -0,0 +1,686 @@
+		{
+			const u16 addr = BIT(op, 2, 14);
+			MCA = addr;
+		}
+
+	}
+}
+
+void kl1839vm1_device::rts(u32 op)
+{
+	MCA = RV;
+}
+
+void kl1839vm1_device::acc(u32 op)
+{
+	UNIMPLEMENTED("ACC");
+}
+
+void kl1839vm1_device::chka(u32 op)
+{
+	const bool madr = BIT(op, 16);
+	if (madr)
+	{
+		const u16 addr_hi = BIT(op, 8, 8) << 6;
+		RC = m_microcode.read_dword(addr_hi | (R(0x18) & 0x3f));
+	}
+	else
+	{
+		const u16 addr = BIT(op, 2, 14);
+		RC= m_microcode.read_dword(addr);
+	}
+}
+
+void kl1839vm1_device::chlk(u32 op)
+{
+	const bool sb = BIT(op, 1);
+	const u32 cnst = BIT(op, 2, 24);
+	RC = (sb * 0xff000000) | cnst;
+}
+
+void kl1839vm1_device::srf(u32 op)
+{
+	if (BIT(op, 25)) // WIMM
+	{
+		UNIMPLEMENTED("SRF: WIMM");
+	}
+	if (BIT(op, 21)) // SFP1
+	{
+		m_fp = true;
+	}
+	if (BIT(op, 20)) // RFP1
+	{
+		m_fp = false;
+	}
+	if (BIT(op, 5)) // DEC
+	{
+		UNIMPLEMENTED("SRF: DEC");
+	}
+	if (BIT(op, 4)) // OCT
+	{
+		--MCA;
+	}
+	if (BIT(op, 2)) // JDZRA
+	{
+		UNIMPLEMENTED("SRF: JDZRA");
+	}
+	if (BIT(op, 1)) // INC
+	{
+		UNIMPLEMENTED("SRF: INC");
+	}
+}
+
+void kl1839vm1_device::invalid(u32 op)
+{
+}
+
+void kl1839vm1_device::kob_process(u8 no, u8 fo, u8 kob, u32 data)
+{
+	switch (kob) {
+		case 0b001: // Data Read
+		case 0b010: // Data Write
+		case 0b011: // Read-Modify-Write
+			m_vma_tmp.d = (no << 30) | (fo << 28) | (kob << 24) | data;
+			break;
+		case 0b100: // Write Accum R(17)
+		case 0b101: // Read Command
+		case 0b110: // Offset Write
+			UNIMPLEMENTED("KOB");
+			break;
+
+		case 0b000: // NOP
+		case 0b111: // Reserve
+		default: break;
+	}
+}
+
+void kl1839vm1_device::decode_op(u32 op)
+{
+	if ((op & 0xc0000000) == 0x00000000) // MA
+	{
+		if ((op & 0xff000000) == 0x01000000)
+			flag(op);
+		else if ((op & 0xff000000) == 0x05000000)
+			rest(op);
+		else
+			ma(op);
+	}
+	else if ((op & 0xc0000000) == 0x40000000) // MB
+	{
+		mb(op);
+	}
+	else if ((op & 0xc0000000) == 0x80000000) // MC
+	{
+		mc(op);
+	}
+	else
+	{
+		if ((op & 0xfc000000) == 0xe0000000)
+			mk(op);
+		else if ((op & 0xe0000000) == 0xc0000000)
+			yp(op);
+		else if ((op & 0xfc000000) == 0xec000000)
+			zsch(op);
+		else if ((op & 0xfc000000) == 0xe4000000)
+			psch(op);
+		else if ((op & 0xfc000000) == 0xf0000000)
+			rts(op);
+		else if ((op & 0xfc000000) == 0xe8000000)
+			acc(op);
+		else if ((op & 0xfc000000) == 0xf4000000)
+			chka(op);
+		else if ((op & 0xfc000000) == 0xf8000000)
+			chlk(op);
+		else if ((op & 0xfc000000) == 0xfc000000)
+			srf(op);
+		else
+			invalid(op);
+	}
+}
+
+void kl1839vm1_device::device_start()
+{
+	space(AS_OPCODES).cache(m_microcode);
+	space(AS_DATA).specific(m_sysram);
+	space(AS_PROGRAM).specific(m_ram);
+	space(AS_IO).specific(m_io);
+
+	save_item(NAME(m_mca));
+	save_item(NAME(m_vma));
+	save_item(NAME(m_vma_tmp));
+	save_item(NAME(m_rv));
+	save_item(NAME(m_sch));
+	save_item(NAME(m_rsp));
+	save_item(NAME(m_ppc));
+	save_item(NAME(m_fp));
+	save_pointer(NAME(m_reg), 0x20);
+	save_pointer(NAME(m_consts), 0x10);
```

Comment:
save_item works on arrays (no idea if it works on PAIR).
save_pointer 2nd param is bytes, not array length.

Reply:
No it isn’t, it’s the element count.

From emu/device.h:
```c++
        /// \brief Register a member of a structure in an array of
        ///   indeterminate for save states
        ///
        /// Registers data to be automatically saved/restored when the
        /// length of the outermost array cannot be automatically determined
        /// at compile time.  Can be used with C arrays of indeterminate
        /// length, pointers, and \c std::unique_ptr instances pointing to
        /// arrays.  Use #save_item if length of the array can be determined
        /// at compile time.
        /// \param value Pointer to the array of structures containing the
        ///   member to be saved/restored.  The #STRUCT_MEMBER macro can be
        ///   used to simplify specifying the \p element and \p valname
        ///   arguments at the same time.
        /// \param [in] element Pointer to the member of the structure to
        ///   save/restore.
        /// \param [in] valname The name of the saved item.  The combination
        ///   of the \p valname and \p index arguments must be unique across
        ///   saved items for a device.
        /// \param [in] count The number of elements in the outermost array.
        /// \param [in] index A numeric value to distinguish between saved
        ///   items with the same name.
        /// \sa save_item
        template <typename ItemType, typename StructType, typename ElementType>
        void ATTR_COLD save_pointer(ItemType &&value, ElementType StructType::*element, const char *valname, u32 count, int index = 0)
        {
                assert(m_save);
                m_save->save_pointer(this, name(), tag(), index, std::forward<ItemType>(value), element, valname, count);
        }
```